### ISSUE TITLE
Window exists but does not display on Ubuntu Linux

### ISSUE THREAD

**Issue by Chickenlord7321** [libtcod/libtcod#156]
I'm neither an experienced programmer nor very good with Linux, so I probably just messed something up when trying to set up libtcod, but here goes...

I'm working on a second-year roguelike project with my team and I've encounter an issue that has me, my team mates, and my professor stumped. We followed the set up instructions from the docs and managed to get a simple window up and running on Mac and Windows, but the window does not show up on my Linux machine (I'm running Ubuntu 24.04).  The code compiles with no errors or warnings, but there is no window. We tried running the following test code:

```c
if (SDL_Init(SDL_INIT_VIDEO) < 0) {
    printf("SDL could not initialize! SDL_Error: %s\n", SDL_GetError());
    return 1;
}

SDL_Window* window = SDL_CreateWindow("My SDL2 Window", 
SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED, 
800, 600, SDL_WINDOW_SHOWN);
if (window == NULL) {
    printf("Window could not be created! SDL_Error: %s\n", SDL_GetError());
    return 1;
}

std::cout << SDL_GetError();
```

...But the print inside the if-block never runs. So the code _thinks_ the window exists. I even tried running a similar test with the deprecated initRoot() and isWindowClosed(), and still the same result. Window exists, but it's not displaying.

This is what it looks like on my teammates' machines:

<img width="1392" alt="libtcod-expected" src="https://github.com/user-attachments/assets/49cafb04-289e-4d4f-9303-5cff8fd00129">

I'm not seeing that at all. 

I have triple-checked and my computer is fully up-to-date. Libtcod and SDL are installed, vcpkg is working, the header files are all included.

Eventually the CS help center suggested it could be a problem with Wayland since SDL recently switched to preferring X11 over Wayland. I went into /etc/gdm3/custom.conf and uncommented Wayland=false, which theoretically should make my machine default to X11. Restarted, ran the code again, still no window.

Now comes the really weird part. I just ran the getting-started hello world from the Python3 libtcod docs, and IT WORKS???? The window shows up!! So now I'm really confused and have no idea why it works with Python, but not with C++. Help.

My environment:

- Ubuntu 24.04 (Linux Debian)
- Windowing System: Wayland
- Compiler: g++ 13.2.0 , gcc 13.2.0
- SDL2 2.31.0
- Libtcod 1.24.0
- vcpkg package management program version 2024-10-18

**Comment by HexDecimal** [comment]
I've heard of Wayland issues before but I'm not very familiar with a Linux environment. Make sure the Vcpkg ports are updated and ensure that it builds SDL2 with Wayland support which should be the default.

Also consider enabling all logging in SDL before initialization:
```c
SDL_LogSetAllPriority(SDL_LOG_PRIORITY_VERBOSE);
```

> Now comes the really weird part. I just ran the getting-started hello world from the Python3 libtcod docs, and IT WORKS???? The window shows up!! So now I'm really confused and have no idea why it works with Python, but not with C++. Help.

The Python version doesn't build SDL2 from Vcpkg. It bundles SDL2 from a Docker image specifically for Linux distributions. This could also be an older version of SDL.

**Comment by HexDecimal** [comment]
I'm assuming you're building libtcod from the Vcpkg port and not building it directly from the repo (or using the release builds). The repo has outdated feature flags which would've disabled Wayland support on SDL2.

If you're using the Vcpkg port then ignore this.

**Comment by Chickenlord7321** [comment]
I checked the vcpkg_installed/vcpkg/status file and it lists sdl2 as an installed package. The default-features are: dbus, ibus, wayland, x11

I added dbus and ibus to the dependencies in vcpkg.json, and now I'm getting the error that ibus does not exist, and vcpkg cannot find it when I attempt to install ibus.

![Screenshot from 2024-11-02 12-51-53](https://github.com/user-attachments/assets/60d4531d-a099-43c9-83a5-57740d719f50)

I installed libtcod using vcpkg yes

**Comment by HexDecimal** [comment]
Please post the text instead of a screenshot in the future.

**Comment by HexDecimal** [comment]
I'm unsure, but you might need the development files for these (ibus, etc) installed locally with the distros package manager for Vcpkg to find them.

**Comment by Chickenlord7321** [comment]
Alright, I will try that

**Comment by Chickenlord7321** [comment]
Oh, apparently ibus is already the newest version (1.5.29-2).

I will have to go read and see why/how vcpkg installs ibus

**Comment by Chickenlord7321** [comment]
Leaving this here in case anyone else comes across the same issue, because I somehow got it fixed:

I removed ibus from the dependencies in vcpkg.json, so now it just includes:
```
  "dependencies": [
    "dbus",
    {
      "name": "libtcod",
      "features": [
        "png",
        "sdl",
        "unicode",
        "zlib"
      ]
    },
    "sdl2",
    "wayland"
  ]
}
```

Running `sudo vcpkg install` then installs everything vcpkg needs, which can take a while (mine took 5.5 min).

The output from Vcpkg recommended installing Xorg dependencies to use X11, which I didn't think was really necessary, but I did it anyway with this command:

`sudo apt install libx11-dev libxft-dev libxext-dev`

Running `cmake CMakeLists.txt` then rebuilt a bunch of packages, but cmake found all its targets and built everything successfully. From there, I just ran `make` and then ran the executable.

Hopefully this helps anyone else who stumbles upon this issue!
